File: token/program-2022/src/extension/transfer_fee.rs
use {
    crate::{
        extension::{AccountType, Extension, ExtensionType},
        pod::*,
    },
    bytemuck::{Pod, Zeroable},
};

/// Transfer fee information
#[repr(C)]
#[derive(Clone, Copy, Debug, Default, PartialEq, Pod, Zeroable)]
pub struct TransferFee {
    /// First epoch where the transfer fee takes effect
    pub epoch: PodU64, // Epoch,
    /// Maximum fee assessed on transfers, expressed as an amount of tokens
    pub maximum_fee: PodU64,
    /// Amount of transfer collected as fees, expressed as basis points of the
    /// transfer amount, ie. increments of 0.01%
    pub transfer_fee_basis_points: PodU16,
}

/// Transfer fee extension data for mints.
#[repr(C)]
#[derive(Clone, Copy, Debug, Default, PartialEq, Pod, Zeroable)]
pub struct MintTransferFee {
    /// Optional authority to set the fee
    pub transfer_fee_config_authority: OptionalNonZeroPubkey,
    /// Withdraw from mint instructions must be signed by this key
    pub withheld_withdraw_authority: OptionalNonZeroPubkey,
    /// Withheld transfer fee tokens that have been moved to the mint for withdrawal
    pub withheld_amount: PodU64,
    /// Older transfer fee, used if the current epoch < new_transfer_fee.epoch
    pub older_transfer_fee: TransferFee,
    /// Newer transfer fee, used if the current epoch >= new_transfer_fee.epoch
    pub newer_transfer_fee: TransferFee,
}
impl Extension for MintTransferFee {
    const TYPE: ExtensionType = ExtensionType::MintTransferFee;
    const ACCOUNT_TYPE: AccountType = AccountType::Mint;
}

/// Transfer fee extension data for accounts.
#[repr(C)]
#[derive(Clone, Copy, Debug, Default, PartialEq, Pod, Zeroable)]
pub struct AccountTransferFee {
    /// Amount withheld during transfers, to be harvested to the mint
    pub withheld_amount: PodU64,
}
impl Extension for AccountTransferFee {
    const TYPE: ExtensionType = ExtensionType::AccountTransferFee;
    const ACCOUNT_TYPE: AccountType = AccountType::Account;
}

#[cfg(test)]
pub(crate) mod test {
    use {super::*, solana_program::pubkey::Pubkey, std::convert::TryFrom};

    pub(crate) fn test_mint_transfer_fee() -> MintTransferFee {
        MintTransferFee {
            transfer_fee_config_authority: OptionalNonZeroPubkey::try_from(Some(Pubkey::new(
                &[10; 32],
            )))
            .unwrap(),
            withheld_withdraw_authority: OptionalNonZeroPubkey::try_from(Some(Pubkey::new(
                &[11; 32],
            )))
            .unwrap(),
            withheld_amount: PodU64::from(u64::MAX),
            older_transfer_fee: TransferFee {
                epoch: PodU64::from(1),
                maximum_fee: PodU64::from(10),
                transfer_fee_basis_points: PodU16::from(100),
            },
            newer_transfer_fee: TransferFee {
                epoch: PodU64::from(100),
                maximum_fee: PodU64::from(5_000),
                transfer_fee_basis_points: PodU16::from(1),
            },
        }
    }
}

File: token/js/src/extensions/transferFee.ts
import { struct, u16 } from '@solana/buffer-layout';
import { publicKey, u64 } from '@solana/buffer-layout-utils';
import { PublicKey } from '@solana/web3.js';
import { Account } from '../state';
import { Mint } from '../state/mint';
import { ExtensionType, getExtensionData } from './extensionType';

export const MAX_FEE_BASIS_POINTS = 10_000;
export const ONE_IN_BASIS_POINTS: BigInt = MAX_FEE_BASIS_POINTS as unknown as BigInt;
/** TransferFeeConfig as stored by the program */
export interface TransferFee {
    /** First epoch where the transfer fee takes effect*/
    epoch: BigInt;
    /**Maximum fee assessed on transfers, expressed as an amount of tokens */
    maximum_fee: BigInt;
    /** Amount of transfer collected as fees, expressed as basis points of the */
    /** transfer amount, ie. increments of 0.01% */
    transfer_fee_basis_points: number;
}

/// Transfer fee extension data for mints.
export interface TransferFeeConfig {
    /// Optional authority to set the fee
    transfer_fee_config_authority: PublicKey;
    /// Withdraw from mint instructions must be signed by this key
    withdraw_withheld_authority: PublicKey;
    /// Withheld transfer fee tokens that have been moved to the mint for withdrawal
    withheld_amount: BigInt;
    /// Older transfer fee, used if the current epoch < new_transfer_fee.epoch
    older_transfer_fee: TransferFee;
    /// Newer transfer fee, used if the current epoch >= new_transfer_fee.epoch
    newer_transfer_fee: TransferFee;
}

/** Buffer layout for de/serializing a mint */
export const TransferFeeLayout = struct<TransferFee>([
    u64('epoch'),
    u64('maximum_fee'),
    u16('transfer_fee_basis_point'),
]);

export const TRANSFER_FEE_AMOUNT_SIZE = TransferFeeLayout.span;

/** Buffer layout for de/serializing a mint */
export const TransferFeeConfigLayout = struct<TransferFeeConfig>([
    publicKey('transfer_fee_config_authority'),
    publicKey('withdraw_withheld_authority'),
    u64('withheld_amount'),
    TransferFeeLayout,
    TransferFeeLayout,
]);

export const TRANSFER_FEE_CONFIG_SIZE = TransferFeeConfigLayout.span;

export function getTransferFeeConfig(mint: Mint): TransferFeeConfig | null {
    const extensionData = getExtensionData(ExtensionType.TransferFeeConfig, mint.tlvData);
    if (extensionData !== null) {
        return TransferFeeConfigLayout.decode(extensionData);
    } else {
        return null;
    }
}

export function getTransferFee(account: Account): TransferFee | null {
    const extensionData = getExtensionData(ExtensionType.TransferFeeConfig, account.tlvData);
    if (extensionData !== null) {
        return TransferFeeLayout.decode(extensionData);
    } else {
        return null;
    }
}
